File: algorithms/cpp/2.add_two_numbers.cpp
// Source:  https://leetcode-cn.com/problems/two-sum/
// Author:  Lianfeng Shen
// Date:    2019-02-15

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *head=NULL, **p=&head;
        int carry=0, sum;
        
        while(l1 != NULL || l2 != NULL || carry != 0){
            sum = getValueAndMoveNext(l1) + getValueAndMoveNext(l2) + carry;
            carry = sum / 10;
            
            ListNode *node = new ListNode(sum % 10);
            *p = node;
            p = &(node->next);
        }
        return head;
    }
private:
    int getValueAndMoveNext(ListNode* &l){
        int val = 0;
        if(l != NULL){
            val = l->val;
            l   = l->next;
        }
        return val;
    }
};

File: algorithms/golang/2.AddTwoNumbers.go
// Source : https://leetcode-cn.com/problems/add-two-numbers/
// Author : Lianfeng Shen
// Date   : 2019-04-25

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    p    := &ListNode{0, nil}
    head := p
    
    carry := 0
    for l1 != nil || l2 != nil || carry != 0 {
        sum := getValueAndMoveNext(&l1) + getValueAndMoveNext(&l2) + carry
        carry = sum / 10 
        
        p.Next = &ListNode{sum % 10, nil}
        p = p.Next
    }
    
    return head.Next
}

func getValueAndMoveNext(p **ListNode) int {
    val := 0
    if *p != nil {
        val = (*p).Val
        *p  = (*p).Next
    }
    return val
}
